File: src/ast.ts
import { Identifier } from "./shared";

export type NumericConstant = {
  kind: "numeric-const";
  value: number;
};

export function NumericConstant(value: number): NumericConstant {
  return { value, kind: "numeric-const" };
}

export type StringConstant = {
  kind: "string-const";
  value: string;
};

export function StringConstant(value: string): StringConstant {
  return { kind: "string-const", value };
}

export type Constant = NumericConstant | StringConstant;

export type Complement = { kind: "complement"; expr: Expr; op: "~" };
export function Complement(expr: Expr): Complement {
  return { kind: "complement", expr, op: "~" };
}

export type UnaryMinus = { kind: "unary-minus"; expr: Expr; op: "-" };
export function UnaryMinus(expr: Expr): UnaryMinus {
  return { kind: "unary-minus", expr, op: "-" };
}

export type LogicalNot = { kind: "logical-not"; expr: Expr; op: "!" };
export function LogicalNot(expr: Expr): LogicalNot {
  return { kind: "logical-not", op: "!", expr };
}

export type UnaryExpr = Complement | UnaryMinus | LogicalNot;

export type RelopName =
  | "less"
  | "less-eq"
  | "greater"
  | "greater-eq"
  | "equal"
  | "not-equal";

export type BinaryOpName =
  | "plus"
  | "divide"
  | "multiply"
  | "subtract"
  | "remainder"
  | "and"
  | "or"
  | RelopName;

export type BinaryExpr = {
  kind: "binary-expr";
  operator: BinaryOpName;
  lhs: Expr;
  rhs: Expr;
};
export function BinaryExpr(
  operator: BinaryOpName,
  lhs: Expr,
  rhs: Expr
): BinaryExpr {
  return { kind: "binary-expr", operator, lhs, rhs };
}

export type Var = {
  kind: "var";
  name: Identifier;
};
export function Var(name: Identifier): Var {
  return { kind: "var", name };
}

export type Assignment = {
  kind: "assignment";
  // Any Expr is allowed for the lhs during parsing. We enforce
  // that it must be an lvalue during semantic analysis.
  lhs: Expr;
  rhs: Expr;
};
export function Assignment(lhs: Expr, rhs: Expr): Assignment {
  return { kind: "assignment", lhs, rhs };
}

export type Conditional = {
  kind: "conditional";
  cond: Expr;
  trueExpr: Expr;
  falseExpr: Expr;
};
export function Conditional(
  cond: Expr,
  trueExpr: Expr,
  falseExpr: Expr
): Conditional {
  return { kind: "conditional", cond, trueExpr, falseExpr };
}

/**
 * The Expr type after parsing but before semantic analysis
 */
export type Expr =
  | Constant
  | UnaryExpr
  | BinaryExpr
  | Var
  | Assignment
  | Conditional;

export type ReturnStmt = {
  kind: "return-stmt";
  expr: Expr;
};
export function ReturnStmt(expr: Expr): ReturnStmt {
  return { kind: "return-stmt", expr };
}

export type ExprStmt = {
  kind: "expr-stmt";
  expr: Expr;
};
export function ExprStmt(expr: Expr): ExprStmt {
  return { kind: "expr-stmt", expr };
}

// A null statement i.e. just a semicolon
export type NullStmt = {
  kind: "null-stmt";
};
export function NullStmt(): NullStmt {
  return { kind: "null-stmt" };
}

export type IfStmt = {
  kind: "if-stmt";
  cond: Expr;
  thenStmt: Stmt;
  elseStmt?: Stmt;
};
export function IfStmt(cond: Expr, thenStmt: Stmt, elseStmt?: Stmt): IfStmt {
  return { kind: "if-stmt", cond, thenStmt, elseStmt };
}

export type Stmt = ReturnStmt | ExprStmt | NullStmt | IfStmt;

export type Declaration = {
  kind: "declaration";
  name: Identifier;
  init: Expr | null;
};
export function Declaration(name: Identifier, init: Expr | null): Declaration {
  const res: Declaration = { kind: "declaration", name, init };
  return res;
}

// Things that can appear at the top-level in a function
export type BlockItem = Stmt | Declaration;

export type Function = {
  name: Identifier;
  body: BlockItem[];
};

export function Function(name: Identifier, body: BlockItem[]): Function {
  return { name, body };
}

export type Program = {
  function_definition: Function;
};

export function Program(function_definition: Function): Program {
  return { function_definition };
}

File: src/ast.rs
//! The notcc AST

use crate::shared_types::Identifier;

// TODO where to put source locations?

#[derive(Clone, PartialEq, Debug)]
/// Unary operator kinds
pub enum UnaryOperator {
  Complement,
  Minus,
  LogicalNot,
}

#[derive(Clone, PartialEq, Debug)]
/// Binary operators
pub enum BinaryOp {
  Plus,
  Divide,
  Multiply,
  Subtract,
  Remainder,
  And,
  Or,
  Less,
  LessEqual,
  Greater,
  GreaterEqual,
  Equal,
  NotEqual,
}

/// Reference type for nested Exprs to break type cycles
pub type ExprRef = Box<Expr>;

#[derive(Clone, PartialEq, Debug)]
// TODO should we put children in an array of Expr for easier handling later?
/// Expression definition including necessary children
pub enum Expr {
  IntConstant(i32),
  // TODO should I use box?
  UnaryExpr(UnaryOperator, ExprRef),
  BinaryExpr(BinaryOp, ExprRef, ExprRef),
  Var(Identifier),
  Assignment(ExprRef, ExprRef),
  Conditional {
    cond: ExprRef,
    true_expr: ExprRef,
    false_expr: ExprRef,
  },
}

impl Expr {
  pub fn kind_to_str(&self) -> &str {
    match self {
      Expr::IntConstant(_) => "Constant",
      Expr::UnaryExpr(_, _) => "UnaryOperator",
      Expr::BinaryExpr(_, _, _) => "BinaryOperator",
      Expr::Var(_) => "Variable",
      Expr::Assignment(_, _) => "Assignment",
      Expr::Conditional {
        cond: _,
        true_expr: _,
        false_expr: _,
      } => "Conditional",
    }
  }
}

#[derive(Clone, PartialEq, Debug)]
/// Statements
pub enum Stmt {
  Return(Expr),
  Expr(Expr),
  Null,
  If {
    cond: Expr,
    true_stmt: Box<Stmt>,
    false_stmt: Option<Box<Stmt>>,
  },
  Compound(Block),
}

#[derive(PartialEq, Debug, Clone)]
/// BlockItem instances can appear at the top level in a function
pub enum BlockItem {
  Declaration(Identifier, Option<Expr>),
  Stmt(Stmt),
}

#[derive(PartialEq, Debug, Clone)]
/// Blocks encompass 0 or more BlockItems
pub struct Block {
  pub items: Vec<BlockItem>,
}

impl Block {
  pub fn new(items: Vec<BlockItem>) -> Block {
    Block { items }
  }
}
#[derive(PartialEq, Debug)]
/// A function's representation
pub struct Function {
  pub name: Identifier,
  pub body: Block,
}

#[derive(PartialEq, Debug)]
/// A program
pub struct Program {
  pub function: Function,
}

#[cfg(test)]
mod tests {
  use super::*;

  #[allow(unused_imports)]
  use pretty_assertions::{assert_eq, assert_ne, assert_str_eq};

  #[test]
  fn construct_ast() {
    // TODO clone is bad. Maybe?
    // TODO better pattern than if..let?
    // TODO What if the if..let doesn't match?
    let num_c = Expr::IntConstant(12);
    let ret_s = Stmt::Return(num_c.clone());
    let func = Function {
      name: Identifier::new("main"),
      body: Block::new(vec![BlockItem::Stmt(ret_s.clone())]),
    };
    assert_eq!(func.name, "main");
    assert_eq!(func.body.items.len(), 1);
    assert_eq!(func.body.items[0], BlockItem::Stmt(ret_s.clone()));

    let decl = BlockItem::Declaration(Identifier::new("var1"), None);
    let decl_init = BlockItem::Declaration(Identifier::new("var1"), Some(num_c.clone()));
    let func2 = Function {
      name: Identifier::new("main"),
      body: Block::new(vec![BlockItem::Stmt(ret_s.clone()), decl, decl_init]),
    };
    let prog = Program { function: func };
    assert!(!std::ptr::eq(&prog.function, &func2));

    let u_minus = Expr::UnaryExpr(UnaryOperator::Minus, Box::new(num_c.clone()));
    assert!(matches!(&u_minus, Expr::UnaryExpr(UnaryOperator::Minus, expr) if **expr == num_c));

    let comp = Expr::UnaryExpr(UnaryOperator::Complement, Box::new(u_minus.clone()));
    assert!(matches!(&comp, Expr::UnaryExpr(UnaryOperator::Complement, expr) if **expr == u_minus));

    let v = Expr::Var(Identifier::new("var1"));
    let assign = Expr::Assignment(Box::new(v.clone()), Box::new(num_c.clone()));
    let expr_s = Stmt::Expr(assign);
    let null_s = Stmt::Null;
    let if_stmt_no_else = Stmt::If {
      cond: num_c.clone(),
      true_stmt: Box::new(expr_s.clone()),
      false_stmt: None,
    };
    assert!(
      matches!(if_stmt_no_else, Stmt::If{cond, true_stmt, false_stmt} if cond == num_c && *true_stmt == expr_s && false_stmt == None )
    );
    let if_stmt_else = Stmt::If {
      cond: v.clone(),
      true_stmt: Box::new(null_s.clone()),
      false_stmt: Some(Box::new(ret_s.clone())),
    };
    assert!(
      matches!(if_stmt_else, Stmt::If{cond, true_stmt, false_stmt} if cond == v && *true_stmt == null_s && **(false_stmt.as_ref().unwrap()) == ret_s )
    );

    let cond_e = Expr::Conditional {
      cond: Box::new(num_c.clone()),
      true_expr: Box::new(v.clone()),
      false_expr: Box::new(u_minus.clone()),
    };
    assert!(
      matches!(cond_e, Expr::Conditional{cond, true_expr, false_expr} if *cond == num_c && *true_expr == v && *false_expr == u_minus)
    );
  }
}
